core/src/forward_packet_batches_by_accounts.rs

```diff
@@ -0,0 +1,335 @@
+use {
+    crate::unprocessed_packet_batches::{self, ImmutableDeserializedPacket},
+    solana_perf::packet::Packet,
+    solana_runtime::{
+        bank::Bank,
+        block_cost_limits,
+        cost_tracker::{CostTracker, CostTrackerError},
+    },
+    solana_sdk::pubkey::Pubkey,
+    std::{rc::Rc, sync::Arc},
+};
+
+/// `ForwardBatch` to have half of default cost_tracker limits, as smaller batch
+/// allows better granularity in interating through buffer, and faster forwarding.
+const DEFAULT_LIMIT_RATIO: u32 = 2;
+/// this number divided by`DEFAULT_LIMIT_RATIO` is the total blocks to forward.
+/// To accommodate transactions without `compute_budget` instruction, which will
+/// have default 200_000 compute units, it has 100 batches as default to forward
+/// up to 12_000 such transaction. (120 such transactions fill up a batch, 100
+/// batches allows 12_000 transactions)
```

Comment:
conversation moved here to original discussion: https://github.com/solana-labs/solana/pull/25406#discussion_r912283531

Reply:
Yea, basically forwarding buckets have capacity for _at least_ 12K transactions, it could batch much more low cost transactions. This const defines _capacity_ of forwarding buckets. Currently all un-forwarded transactions in buffer are sorted into account buckets, but number of transactions being forwarded are limited by data_budget (same as before).

In the scenario that there are 20K un-forwarded transactions in buffer, knowing there is 10K/sec limit for threads, in order to avoid sorting all 20K txs, I think can put logic in `banking_stage::filter_valid_packets_for_forwarding` when buffer is iterated. I also think this can be a separate PR as enhancement. wdyt?

Reply:
yeah i think `filter_valid_packets_for_forwarding` is a good place to truncate the batch. Separate PR sounds good to me

Reply:
#26433